boring/src/aead.rs

```diff
@@ -0,0 +1,357 @@
+                output_len,
+                nonce.as_ptr(),
+                nonce_len,
+                data.as_ptr(),
+                data_len,
+                aad.as_ptr(),
+                aad_len,
+            ))?;
+
+            Ok(output_len as usize)
+        }
+    }
+
+    /// Opens data by authenticating the provided aad and authenticating and decrypting the provided data.
+    ///
+    /// Returns the number of bytes written to output.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `output.len() > c_int::max_value()`.
+    ///
+    /// Panics if `output.len() < data.len()`.
+    pub fn open(
+        &mut self,
+        nonce: &[u8],
+        aad: &[u8],
+        data: &[u8],
+        output: &mut [u8],
+    ) -> Result<usize, ErrorStack> {
+        unsafe {
+            assert!(output.len() <= c_int::max_value() as usize);
+            assert!(output.len() >= data.len());
+
+            let mut output_len = output.len();
+            let nonce_len = nonce.len();
+            let data_len = data.len();
+            let aad_len = aad.len();
+
+            cvt(ffi::EVP_AEAD_CTX_open(
+                self.ctx,
+                output.as_mut_ptr(),
+                &mut output_len,
+                output_len,
+                nonce.as_ptr(),
+                nonce_len,
+                data.as_ptr(),
+                data_len,
+                aad.as_ptr(),
+                aad_len,
+            ))?;
+
+            Ok(output_len)
+        }
+    }
+}
+
+impl Drop for AeadCrypter {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::EVP_AEAD_CTX_free(self.ctx);
+        }
+    }
+}
+
+/// Encrypts the provided data and authenticates the ciphertext and the provided aad.
+///
+/// Returns the sealed data as a vector.
+pub fn encrypt_aead(
+    aead: Aead,
+    key: &[u8],
+    nonce: &[u8],
+    aad: &[u8],
+    data: &[u8],
+) -> Result<Vec<u8>, ErrorStack> {
+    let mut aeadcrypter = AeadCrypter::new(aead, key)?;
+    let mut output = vec![0u8; data.len() + aead.max_ciphertext_overhead()];
+    let bytes_written_to_output = aeadcrypter.seal(nonce, aad, data, &mut output)?;
+    Ok(output[..bytes_written_to_output].to_vec())
+}
+
+/// Authenticates the provided data and aad and decrypts the provided data.
+///
+/// Returns the unsealed data as a vector.
+pub fn decrypt_aead(
+    aead: Aead,
+    key: &[u8],
+    nonce: &[u8],
+    aad: &[u8],
+    data: &[u8],
+) -> Result<Vec<u8>, ErrorStack> {
+    let mut aeadcrypter = AeadCrypter::new(aead, key)?;
+    let mut output = vec![0u8; data.len()];
+    let bytes_written_to_output = aeadcrypter.open(nonce, aad, data, &mut output)?;
+    Ok(output[..bytes_written_to_output].to_vec())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hex::{self, FromHex};
+
+    #[test]
+    fn test_aead_encrypt_decrypt() {
+        let key = [0u8; 16];
+        let nonce = [0u8; 16];
+        let aad = [0u8; 16];
+        let data = [0u8; 16];
```

Comment:
Could you use something other than zeros here? I'm not sure what boringssl does on zeros, and it seems better to be closer to a real scenario anyway.